src/Bicep.LangServer.IntegrationTests/CompletionTests.cs

```diff
@@ -51,18 +52,94 @@ public async Task EmptyFileShouldProduceDeclarationCompletions()
             if (expectedStr == null)
             {
                 throw new AssertFailedException($"The completion set '{expectedSetName}' does not exist.");
             }
-            
+
             var expected = JToken.Parse(expectedStr);
 
             actual.Should().EqualWithJsonDiffOutput(TestContext, expected, GetGlobalCompletionSetPath(expectedSetName), actualLocation);
         }
 
+        [TestMethod]
+        public async Task ValidateSnippetCompletionAfterPlaceholderReplacements()
+        {
+            Dictionary<Uri, string> fileSystemDict = new Dictionary<Uri, string>();
+            MultipleMessageListener<PublishDiagnosticsParams> diagnosticsListener = new MultipleMessageListener<PublishDiagnosticsParams>();
+
+            ILanguageClient client = await IntegrationTestHelper.StartServerWithClientConnectionAsync(
+                options =>
+                {
+                    options.OnPublishDiagnostics(diags => diagnosticsListener.AddMessage(diags));
+                },
+                fileResolver: new InMemoryFileResolver(fileSystemDict));
+
+            string? currentDirectory = Path.GetDirectoryName(Assembly.GetEntryAssembly()?.Location);
+            string snippetTemplatesFolder = Path.Combine(currentDirectory!, "Completions", "SnippetTemplates");
+
+            IEnumerable<CompletionItem> snippetCompletions = await GetDeclationSnippetCompletionsAsync();
+
+            foreach (CompletionItem completionItem in snippetCompletions)
+            {
+                string placeholderFile = Path.Combine(snippetTemplatesFolder, completionItem.Label + ".bicep");
+                DocumentUri documentUri = DocumentUri.FromFileSystemPath(placeholderFile);
+                string bicepFileWithPlaceholderReplacementsAndErrorCodes = File.ReadAllText(placeholderFile);
+
+                string snippetTextAfterReplacements = SnippetCompletionTestHelper.GetSnippetTextAfterPlaceholderReplacements(completionItem, bicepFileWithPlaceholderReplacementsAndErrorCodes);
+                fileSystemDict[documentUri.ToUri()] = bicepFileWithPlaceholderReplacementsAndErrorCodes.Replace("// Insert snippet here", snippetTextAfterReplacements);
+
+                client.TextDocument.DidOpenTextDocument(TextDocumentParamHelper.CreateDidOpenDocumentParams(documentUri, fileSystemDict[documentUri.ToUri()], 1));
+
+                var diagsParams = await diagnosticsListener.WaitNext();
+                diagsParams.Uri.Should().Be(documentUri);
+
+                List<string> expectedErrorCodes = SnippetCompletionTestHelper.GetErrorCodes(completionItem.Detail!, bicepFileWithPlaceholderReplacementsAndErrorCodes);
+
+                Container<Diagnostic> diagnostics = diagsParams.Diagnostics;
+
+                // We need to make sure that actual errors/warnings match error codes documented in test bicep file
+                if (!expectedErrorCodes.Any())
```

Comment:
What sorts of expected errors are there?

Reply:
This is for the ones that we deprecated:
![image](https://user-images.githubusercontent.com/30270536/112376738-0a47cb80-8ca2-11eb-9c69-64949f59a5f0.png)

Reply:
Ah - so for the param syntax in particular, I think our snippets should be up-to-date with the latest language spec. So rather than offering:
```
param Identifier Type {
  default: $default
  allowed: [
    $allowed
  ]
}
```

We should be offering:
```
@allowed([
  $allowed
])
param Identifier Type = $default
```

Generally, I think we should just be strict with snippets + snippet templates from the start, and just say "no warnings or errors". It means that someone can't get away with checking in a new language feature that breaks snippets, and it means that someone can't get away with checking in a snippet that doesn't actually work.

This is what we do with examples; I don't think it's been a barrier for user contributions, and it's really nice to have the confidence that if someone pulls down one of our examples, they're not going to see errors/warnings.

Reply:
Yeah, it's a deprecated feature, so we definitely shouldn't encourage its use.

---
In reply to: [600905308](https://github.com/Azure/bicep/pull/1999#discussion_r600905308) [](ancestors = 600905308)

Reply:
Btw, I agree in principle that we shouldn't have expected errors in snippers since we have replacement values for each one. However, we may run into cases where a swagger/types bug causes that to happen but it's still valueable to have the ability to opt out. Since @bhsubra already implemented it, I think we should leave the capability in but just not have any expected errors in the current set of snippets.

---
In reply to: [600923276](https://github.com/Azure/bicep/pull/1999#discussion_r600923276) [](ancestors = 600923276,600905308)